src/main/java/com/microsoft/aad/msal4j/TokenCache.java

```diff
@@ -300,16 +302,92 @@ static private AppMetadataCacheEntity createAppMetadataCacheEntity(TokenRequestE
                         build())) {
             try {
                 lock.readLock().lock();
+                Set<IAccount> rootAccounts = new HashSet<>();
 
-                return accounts.values().stream().
+                //Filter accounts map into two sets: a set of home accounts (local OID = home OID),
+                // and a set of tenant profiles (local OID != home OID)
+                Set<AccountCacheEntity> homeAccounts = accounts.values().stream().
                         filter(acc -> environmentAliases.contains(acc.environment())).
-                        collect(Collectors.mapping(AccountCacheEntity::toAccount, Collectors.toSet()));
+                        filter(acc -> acc.homeAccountId().contains(acc.localAccountId())).
+                        collect(Collectors.toSet());
+                Set<AccountCacheEntity> tenantAccounts = accounts.values().stream().
+                        filter(acc -> environmentAliases.contains(acc.environment())).
+                        filter(acc -> !acc.homeAccountId().contains(acc.localAccountId())).
+                        collect(Collectors.toSet());
+
+                for (AccountCacheEntity homeAccCached : homeAccounts) {
+                    IdTokenCacheEntity homeAcctIdToken = idTokens.get(getIdTokenKey(
```

Comment:
should not normally happen , but it is good to make sure it exist and not null

Reply:
Good call, I'll add this check in with the next commit with your other recommendations.